src/pkcs11_uri_wrapped.rs

```diff
@@ -0,0 +1,229 @@
+                    None => return Err(anyhow!("")),
+                };
+                match p {
+                    "" | "file" => {
+                        if !std::path::Path::new(pinuri.path()).is_absolute() {
+                            return Err(anyhow!("PIN URI path {} is not absolute", pinuri.path()));
+                        }
+                        let pin = std::fs::read_to_string(pinuri.path())?;
+                        return Ok(pin)
+                    },
+                    _ => {
+                        let ss = pinuri.scheme_str();
+                        match ss {
+                            Some(s) => return Err(anyhow!("PIN URI scheme {} is not supported", s)),
+                            None => return Err(anyhow!("failed to get scheme from pin URI")),
+                        }
+                    },
+                }
+            },
+            None => {},
+        }
+        // FIXME error
+        Err(anyhow!("Neither pin-source nor pin-value are available"))
+    }
+
+
+    // has_pin allows the user to check whether a PIN has been provided either by
+    // the pin-value or the pin-source attributes. It should be called before
+    // pin(), which may still fail getting the PIN from a file for example.
+    pub fn has_pin(&self) -> bool {
+        match &self.p11uri.query_attributes.pin_value {
+            Some(_x) => return true,
+            None => {},
+        }
+        match &self.p11uri.query_attributes.pin_source {
+            Some(_x) => return true,
+            None => {},
+        }
+        false
+    }
+
+
+    // Get the module to use or an error in case no module could be found.
+    // First the module-path is checked for whether it holds an absolute that
+    // can be read by the current user. If this is the case the module is
+    // returned. Otherwise either the module-path is used or the user-provided
+    // module path is used to match a module containing what is set in the
+    // attribute module-name.
+    pub fn module(&self) -> Result<String> {
+        let searchdirs_tmp: Vec<String>; // FIXME
+        let searchdirs: &Vec<String>;
+
+        if let Some(mp) = &self.p11uri.query_attributes.module_path {
+            let info = std::fs::metadata(mp)?;
+            if info.is_file() {
+                // it's a file
+                if self.is_allowed_path(&mp, &self.allowed_module_paths)? {
+                    return Ok(mp.to_string());
+                }
+                return Err(anyhow!(""));
+            }
+            if !info.is_dir() {
+                // it's a symlink
+                return Err(anyhow!(""));
+            }
+            // it's a dir
+            searchdirs_tmp = vec![mp.to_string()];
+            searchdirs = &searchdirs_tmp;
+        } else {
+            searchdirs = &self.module_directories;
+        }
+
+        let module_name = match &self.p11uri.query_attributes.module_name {
+            Some(mn) => mn.to_lowercase(),
```

Comment:
not sure `to_lowercase` is a good approach here (and below) -- linux filesystem is case sensitive; you can have two files in a directory called "MODule_1" and "module_1".